qiskit/algorithms/optimizers/suffix_averaging.py

```diff
@@ -0,0 +1,154 @@
+    obtained from :math:`T` iterations, :math:`\{\vec{\theta}^{(t)}\}_{t=1}^{t=T}`,
+    the SuffixAveragingOptimizer returns the averaged point of the last n_params_suffix
+    points,
+
+    .. math::
+
+        \overline{\vec{\theta}} = \frac{1}{n_params_suffix} \sum_{t=T-n_params_suffix-1}^{T}
+        \vec{\theta}^{(t)}.
+
+    Examples:
+        .. code-block::python
+            from qiskit import Aer
+            from qiskit.utils import QuantumInstance
+            from qiskit.algorithms import VQE
+            from qiskit.opflow.gradients import Gradient
+            from qiskit.circuit.library import TwoLocal
+            from qiskit.algorithms.optimizers import ADAM
+            from qiskit.algorithms.optimizers.suffix_averaging import SuffixAveragingOptimizer
+
+            backend = Aer.get_backend("qasm_simulator")
+            quantum_instance = QuantumInstance(backend=backend, shots=1000)
+            ansatz = TwoLocal(4, ["rx", "rz"], "cx", "linear")
+            optimizer = ADAM(maxiter=1000)
+            suffix_optimizer = SuffixAveragingOptimizer(optimizer, n_params_suffix=50)
+            gradient = Gradient(grad_method="param_shift")
+            vqe = VQE(
+                ansatz = ansatz,
+                optimizer = suffix_optimizer,
+                gradient = gradient,
+                quantum_instance = quantum_instance
+            )
+
+    References:
+        [1] S. Tamiya and H. Yamasaki. 2021.
+        Stochastic Gradient Line Bayesian Optimization
+        for Efficient Noise-Robust Optimization of Parameterized Quantum Circuits.
+        arXiv preprint arXiv:2111.07952.
+    """
+
+    def __init__(self, optimizer: Optimizer, n_params_suffix: int = 50) -> None:
+        """
+        Args:
+            optimizer: The optimizer used for optimizing parameterized quantum circuits.
+            n_params_suffix: The number of circuit parameters for taking the suffix average.
+        """
+
+        self._n_params_suffix = n_params_suffix
+        self._optimizer = optimizer
+
+        self._circ_params = []
+
+        original_callback = self._optimizer.callback
+        if isinstance(self._optimizer, (SPSA, QNSPSA)):
+
+            def callback(nfev, x_next, fx_next, update_step, is_accepted):
+                self._circ_params.append(x_next)
+                if original_callback is not None:
+                    original_callback(nfev, x_next, fx_next, update_step, is_accepted)
+
+        elif isinstance(self._optimizer, GradientDescent):
+
+            def callback(nfevs, x_next, fx_next, stepsize):
+                self._circ_params.append(x_next)
+                if original_callback is not None:
+                    original_callback(nfevs, x_next, fx_next, stepsize)
+
+        else:
+
+            def callback(x):
+                self._circ_params.append(x)
+                if original_callback is not None:
+                    original_callback(x)
+
+        self._optimizer.callback = callback
```

Comment:
There are other optimizers, including the nlopts, that do not have any callback support. Setting a callback on these will do nothing and presumably this optimizer will fail to work as expected. Should we have some indication as to whether callback is supported or not so this can check and raise an error accordingly?

Reply:
In this case, I think callback setter in npopts should raise an error.

Reply:
Since this setting in done in the constructor that could be reasonable as it would fail when the suffix_averaging optimizer was built. It snot just the nlopt optimizers in that folder there are others in the optimizers such as bobyqa, imfil, snobfit. Maybe we could arrange for the default implementation in base class to raise an error and then override it say in SciPyOptimzer, (QN)SPSA etc where it is supported since these classes are being modified anyway so that the default, by virtue of the subclass, is that its not implemented, and we override where it is.

Reply:
It sounds nice. I'll try it.

Reply:
Could this section be simplified by something like
```python
def callback(x, *args, **kwargs):
    self._circ_params.append(x)
    original_callback(x, *args, **kwargs)
```
?

Reply:
No, because x is not always the first. (It's complicated because various callbacks already exist. I understand that the `scipy.optimize.minimize` one alone is not user friendly, but...)

Reply:
Hmm.. in that case do we need to add the `OptimizerCallback` at all? They differ a lot anyways and the `SuffixAveraging` needs to handle the different optimizers anyways and cannot really leverage the signature from `OptimizerCallback`?

In future we could maybe add it by unifying the callback signature to make sure the point always comes first.

Reply:
Yes, I think so. Personally, I would like to unify them.

```python
def __call__(self, xk, state: Optional[OptimizeResult]) -> Optional[bool]:
```